### PR TITLE
Hermetic Python 3 runtime by building from source + jupyter notebook

### LINKED ISSUE #1817: Create a Python environment separate from the system interpreter and use in in Bazel
### Description of the task

Create a Python environment separate from the system interpreter and use in in Bazel. 

### Acceptance criteria

- [ ] Script to setup Python environment
- [ ] Make sure Bazel can find and use the environment

### LINKED ISSUE #1818: Launch Jupyter Notebook with pyenv-virtualenv Python environment as a Bazel target
### Description of the task

Launch Jupyter Notebook with pyenv-virtualenv Python environment as a Bazel target

### Acceptance criteria

- [ ] Launch Jupyter Notebook with pyenv-virtualenv Python environment as a Bazel target

### Blocked By

#1817

### PR DESCRIPTION
**Pull request by jordanzliu**
### Description

Hermetic Python 3 runtime by building from source

### Testing Done

Added a Python test to ensure that `/usr/bin/python*` isn't being used as the interpreter
All existing tests pass

### Resolved Issues

resolves #1817, resolves #1818 

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [ ] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [ ] **Remove all commented out code**
- [ ] **Remove extra print statements**: for example, those just used for testing
- [ ] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue
- [ ] **Justify drops in code coverage**: If this PR results in a non-trivial drop in code coverage (a bot should post a coverage diagram as a comment), please justify why we can't test the code that's not covered.

### CODE REVIEW

File: environment_setup/setup_software.sh [hunk 1]
```diff
@@ -58,6 +58,7 @@ host_software_packages=(
     python3-yaml # Load dynamic parameter configuration files
     qt5-default # The GUI library for our visualizer
     valgrind # Checks for memory leaks
+    libsqlite3-dev # needed to build Python 3 with sqlite support
```
Comment (MathewMacDougall) [inline environment_setup/setup_software.sh:61]: Do we need sqlite support? Db for ML/proto data?

**Review by MathewMacDougall — CHANGES_REQUESTED** [review]
This is actually really clean. Nice work.

Now that we have a python toolchain that ideally everything should use, I don't suppose there's any way to enforce it? ie. warn if a python rule doesn't specify the interpreter.

Also should we switch the dynamic paramters generation scripts to use this too?

File: environment_setup/setup_software.sh (hunk 1 above)
Comment (jordanzliu) [reply environment_setup/setup_software.sh:61]: Having the built-in sqlite is a dependency of Jupyter Notebook

**Comment by jordanzliu** [comment]
@MathewMacDougall parameters v1 explicitly calls the system interpreter, and parameters v2 uses `py_binary` which uses the declared Python toolchain. All Bazel Python targets will uses the declared toolchain. 

Update: tried using `execpath` to use the hermetic interpreter but there's no way to pull in `pyyaml` as a dependency with a `genrule`. Leaving it for now.

**Review by MathewMacDougall — DISMISSED** [review]
LGTM

@garethellis0 please take a look too

File: src/WORKSPACE [hunk 2]
```diff
@@ -183,35 +183,74 @@ http_archive(
...
-load("@rules_python//python:repositories.bzl", "py_repositories")
+# Based on the hermetic Python 3 guide found here
+# https://thethoughtfulkoala.com/posts/2020/05/16/bazel-hermetic-python.html
 
-py_repositories()
+# Special logic for building python interpreter with OpenSSL from homebrew.
+# See https://devguide.python.org/setup/#macos-and-os-x
+_py_configure = """
+if [[ "$OSTYPE" == "darwin"* ]]; then
+    ./configure --prefix=$(pwd)/bazel_install --with-openssl=$(brew --prefix openssl)
```
Comment (akhilveeraghanta) [inline src/WORKSPACE]: we probably won't support `macOS`, do we need this check here?

File: src/python_tools/BUILD [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+)
+
+py_binary(
+    name = "jupyter",
+    srcs = ["jupyter.py"],
+    deps = [
+        requirement("jupyter"),
+        requirement("notebook"),
+    ],
+)
```
Comment (akhilveeraghanta) [inline src/python_tools/BUILD:17]: should `python_tools` be its own folder in the top level directory? Thoughts on having this under `//software/tools:jupyter_notebook`?

**Review by akhilveeraghanta — COMMENTED** [review]
This is awesome, nice work on getting the hermetic python stuff going!

Currently the notebooks are saved to a "local dir" which happens to be: 

~~~
Serving notebooks from local directory: /home/akhil/.cache/bazel/_bazel_akhil/7c62ef8c97a53afa53e3884314b45b1e/execroot/__main__/bazel-out/k8-fastbuild/bin/python_tools/jupyter.runfiles/__main__
~~~

Is there any way we can retarget this to a git ignored folder in the software repo (or something like that) so its not in this weird spot?

File: src/WORKSPACE (hunk 2 above)
Comment (jordanzliu) [reply src/WORKSPACE]: that was copypasted from the hermetic python guide, probably unnecessary

File: src/python_tools/BUILD (hunk 3 above)
Comment (jordanzliu) [reply src/python_tools/BUILD:17]: my idea for this was that `software` would be C++ land and all the python data analysis stuff would go into `python_tools`

**Comment by jordanzliu** [comment]
@akhilveeraghanta the jupyter script is pretty much there just to demonstrate that it's functional, there's followup commits after this at https://github.com/jordanzliu/Software/tree/proto_wrapper that address this with a `notebooks` directory in `python_tools`. The plan is to merge those in the next PR with the protobuf build for Python and wrapper classes for replay data.

**Review by jsn0716 — DISMISSED** [review]
LGTM

**Review by kelvinkoon — DISMISSED** [review]

**Review by jonathanlew — COMMENTED** [review]
Nothing to add, lgtm

**Review by MathewMacDougall — APPROVED** [review]

**Review by kelvinkoon — APPROVED** [review]

**Review by akhilveeraghanta — APPROVED** [review]

**Pull request merged**